### PR TITLE
Implement server

### PR DESCRIPTION
**Pull request by TheNeikos**

### CODE REVIEW

**Comment by TheNeikos** [comment]
The failing check is due to unused parts in the repository. I think they are fine to add as `cargo test` and co report success.

File: mqtt-format/src/v3/packet.rs [hunk 1]
```diff
@@ -214,13 +216,37 @@ impl<'message> MPacket<'message> {
...
+                qos,
+                retain,
+                topic_name,
+                id,
+                payload,
+            } => {
+                let packet_type = 0b0011_0000;
+                let dup_mask = if *dup { 0b0000_1000 } else { 0 };
+                let qos_mask = qos.to_byte() << 1;
+                let retain_mask = *retain as u8;
```
Comment (matthiasbeyer) [inline mqtt-format/src/v3/packet.rs:229]: Some comment here would be nice, as this is very "magic constant"y.

File: src/server.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+    }
+}
+
+impl<'message> TryFrom<MString<'message>> for ClientId {
+    type Error = ClientError;
+
+    fn try_from(ms: MString<'message>) -> Result<Self, Self::Error> {
+        Ok(ClientId(ms.to_string()))
+    }
+}
```
Comment (matthiasbeyer) [inline src/server.rs:90]: Does this actually have to be a `TryFrom`? Could it be a `From`?

File: src/server.rs [hunk 3]
```diff
@@ -0,0 +1,374 @@
...
+                            // 4. If QoS == 2 -> Send into writer && Store Message waiting for PubRec
+                        }
+                        None => {
+                            debug!(
+                                ?publisher_client_id,
+                                "No more senders, stopping sending cycle"
+                            );
+                        }
+                    }
+                }
```
Comment (matthiasbeyer) [inline src/server.rs:248]: Am I missing something or is there a `break` missing somewhere in here?

File: src/server.rs [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+            });
+
+            let keep_alive = *keep_alive;
+            let subscription_manager = self.subscription_manager.clone();
+
+            tokio::spawn(async move {
+                let client_id = client_id;
+                let client_connection = client_connection;
+                let mut reader = client_connection.reader.lock().await;
+                let keep_alive_duration = Duration::from_secs((keep_alive as u64 * 150) / 100);
```
Comment (matthiasbeyer) [inline src/server.rs:258]: Comment on the magic constants here would be nice as well.

**Review by matthiasbeyer — CHANGES_REQUESTED** [review]
Mostly minor things :wink:

File: src/server.rs (hunk 2 above)
Comment (TheNeikos) [reply src/server.rs:90]: Right now yes, but in the future this might fail due to an invalid clientid

File: mqtt-format/src/v3/packet.rs (hunk 1 above)
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:229]: I think the variable names are very clear?

**Comment by TheNeikos** [comment]
So, currently the PR doesn't build due to multiple warnings.

Some of which can be addressed, but several require more in depth design. @matthiasbeyer would you accept that they are fixed in a later PR or do you think one mega PR is better?

**Comment by matthiasbeyer** [comment]
Well, half of them can easily be fixed! Feel free to pull from https://github.com/matthiasbeyer/cloudmqtt/tree/feature/implement_server for fixup patches.

The other half is unused functions or enum variants, which I would say could be `#[allow(...)]`d and be fixed later on.

File: mqtt-format/src/v3/packet.rs [hunk 5]
```diff
@@ -26,62 +27,179 @@ use super::{
     MSResult,
 };
 
+#[cfg_attr(feature = "yoke", derive(yoke::Yokeable))]
```
Comment (matthiasbeyer) [inline mqtt-format/src/v3/packet.rs:30]: This feature does not exist in `Cargo.toml` of the crate.

File: mqtt-format/src/v3/packet.rs [hunk 6]
```diff
@@ -26,62 +27,179 @@ use super::{
...
+            type Error = ();
+
+            fn try_from(value: &'other MPacket<'message>) -> Result<Self, Self::Error> {
+                if let MPacket::$var(var) = value {
+                    Ok(var)
+                } else {
+                    Err(())
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:167]: Because it can fail! Into/From never fails

File: mqtt-format/src/v3/packet.rs (hunk 5 above)
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:30]: Added a note!

File: mqtt-format/src/v3/packet.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (matthiasbeyer) [reply mqtt-format/src/v3/packet.rs:167]: Yes sure, but why does it have to? This code looks like a no-op that except that it wraps the value in a Result?

[the remark this answers is not in the document]
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:167]: It's not a No-op? It goes from `Foo` to `Foo::<Y>` and return `Y` ?

[the remark this answers is not in the document]
Comment (matthiasbeyer) [reply mqtt-format/src/v3/packet.rs:167]: So it is essentially an `unwrap()`?

[the remark this answers is not in the document]
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:167]: No, I don't know where you're going haha

[the remark this answers is not in the document]
Comment (matthiasbeyer) [reply mqtt-format/src/v3/packet.rs:167]: My question is essentially: Can't this be done at compiletime?

[the remark this answers is not in the document]
Comment (TheNeikos) [reply mqtt-format/src/v3/packet.rs:167]: No. `MPacket` could be any of its variants. Hence the potential for failure.

**Review by matthiasbeyer — APPROVED** [review]
The stuff I annotated was fixed. So no point in blocking this! I didn't do a in-depth review though!

**Comment by TheNeikos** [comment]
The in-depth stuff will happen once we got an alpha ;D

**Comment by TheNeikos** [comment]
bors merge

**Pull request merged**